audit-log/README.md

```diff
@@ -0,0 +1,151 @@
+> when it has occurred.
+
+In plain words
+
+> Whenever a some property changes, log the change in a file.
+
+Wikipedia says
+
+> An audit trail (also called audit log) is a security-relevant chronological record, set of 
+> records, and/or destination and source of records that provide documentary evidence of the 
+> sequence of activities that have affected at any time a specific operation, procedure, event, 
+> or device.  
+
+**Programmatic Example**
+In this example implementation, `Customer` is an object that is being logged by an Audit Log, 
+`AuditLog`, and changes to the `Customer`'s address, or name will be logged by the logger.
+
+Here are the relevant parts of `Customer`,
+
+```java
+public class Customer {
+  private String address;
+  private String name;
+  private final int id;
+
+  public Customer(String name, int id) {
+    this.name = name;
+    this.id = id;
+  }
+
+  public void setAddress(String address, SimpleDate changeDate) {
+    AuditLog.log(changeDate, this, "change of address", this.address, address);
+    this.address = address;
+  }
+
+  public void setName(String name, SimpleDate changeDate) {
+    AuditLog.log(changeDate, this, "change of name", this.name, name);
+    this.name = name;
+  }
+}
+```
+
+And here are the relevant parts of `AuditLog`
+
+```java
+public class AuditLog {
+  private static File logFile;
+
+  public static void write(String string) throws IOException {
+    Files.write(getLogFile().toPath(), string.getBytes(StandardCharsets.UTF_8),
+            StandardOpenOption.APPEND);
+  }
+
+  public static void log(SimpleDate date, Customer customer, String description, Object oldValue,
+                         Object newValue) {
+    try {
+      write(SimpleDate.getToday().toString() + "\t" + customer.getId() + "\t" + customer.getName()
+              + "\t" + description + "\t" + oldValue + "\t" + newValue + "\t"
+              + date.toString() + "\n");
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+
+  }
+}
+```
+
+Then, with an empty log file, we have an application using the log.
+```java
+  public static void main(String[] args) {
+    SimpleDate.setToday(new SimpleDate(2000, 5, 7));
+    Customer john = new Customer("John Smith", 1);
+
+    john.setAddress("1234 Street St", new SimpleDate(1999, 1, 4));
+
+    // change dates to simulate time passing
+    SimpleDate.setToday(new SimpleDate(2001, 6, 3));
+
+    // note that john's address and name have changed in the past
+    john.setName("John Johnson", new SimpleDate(2001, 4, 17));
+    john.setAddress("4321 House Ave", new SimpleDate(2000, 8, 23));
+  }
+```
+
+Then, the log file after this has been run is 
+```
+2000, 5, 7	1	John Smith	change of address	null	1234 Street St	1999, 1, 4
+2001, 6, 3	1	John Smith	change of name	John Smith	John Johnson	2001, 4, 17
+2001, 6, 3	1	John Johnson	change of address	1234 Street St	4321 House Ave	2000, 8, 23
```

Comment:
This formatting in this output is a bit hard to read. If it is a simple chance, I recommend making the `John Smith` line up with the other elements in its column.